x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/ingest/AnalyticsEventEmitter.java

```diff
@@ -75,11 +96,23 @@ public void emitEvent(
         } catch (IOException e) {
             listener.onFailure(new ElasticsearchException("Unable to parse the event.", e));
         } catch (EsRejectedExecutionException e) {
-            listener.onFailure(new ElasticsearchException("Unable to add the event to the bulk."));
-            logger.error("Unable to add the event to the bulk.", e);
+            onEventDrop(listener);
         }
     }
 
+    private void onEventDrop(final ActionListener<PostAnalyticsEventAction.Response> listener) {
+        listener.onFailure(new ElasticsearchStatusException("Unable to add the event: too many requests.", RestStatus.TOO_MANY_REQUESTS));
+        if (dropEvent.compareAndSet(false, true)) {
+            logger.warn(Strings.format("Bulk processor is full. Start dropping events for %s.", coolDownDelay.getStringRep()));
+            client.threadPool().schedule(this::onCoolDownExpired, coolDownDelay, ThreadPool.Names.SAME);
```

Comment:
Do we need thread scheduling for cooling down? Wouldn't it be equally effective to register the time at which we can accept events again, and compare the current time with it?

Reply:
I like the idea but I am still struggling to know how to log a message to notify that we are accepting event again...
Any suggestions?

PS: I think it is important at least until we will have better monitoring for event ingest in 8.9

Reply:
have you considered using BulkProcessor's backpressure mechanism and retry features?
I think it should give use the same "drop event" functionality

perhaps just using bulkdProcessor.addWithBackpressure(eventIndexRequest, ()->false) would be enough?

Reply:
@pgomulka I wanted to use it first but the implementation adds wait time of 500ms. In the context of handling a lot of parallel HTTP connections this  block could be a problem. I could do it async but would loose the ability to respond with a 429 error and the risk of having this monitored. 

Instead I made the choice of not trying to add new events until the cool down delay is expired. 

More specifically, this cool down mechanism helped me to deal with OOM that I was able to see on my laptop:

```
One instance of org.elasticsearch.index.translog.Translog loaded by jdk.internal.loader.ClassLoaders$AppClassLoader @ 0x7e00008b0 occupies 99 020 568 (39,68 %) bytes. The memory is accumulated in one instance of java.util.HashMap$Node[], loaded by <system class loader>, which occupies 98 565 752 (39,50 %) bytes.
```

Reply:
Also I did new Rally tests with all the optimizations I made since I introduce it and this cool down mechanism does not seems to be required anymore.

Reply:
I can confirm this is not needed anymore. Reverted the code to a previous which is just logging start / end of event dropping




```java
    public void emitEvent(
        final PostAnalyticsEventAction.Request request,
        final ActionListener<PostAnalyticsEventAction.Response> listener
    ) {
        try {
            AnalyticsEvent event = eventFactory.fromRequest(request);
            IndexRequest eventIndexRequest = createIndexRequest(event);

            bulkProcessor.add(eventIndexRequest);

            if (dropEvent.compareAndSet(true, false)) {
                logger.warn("Bulk processor has been flushed. Accepting new events again.");
            }

            if (request.isDebug()) {
                listener.onResponse(new PostAnalyticsEventAction.DebugResponse(true, event));
            } else {
                listener.onResponse(PostAnalyticsEventAction.Response.ACCEPTED);
            }
        } catch (IOException e) {
            listener.onFailure(new ElasticsearchException("Unable to parse the event.", e));
        } catch (EsRejectedExecutionException e) {
            listener.onFailure(
                new ElasticsearchStatusException("Unable to add the event: too many requests.", RestStatus.TOO_MANY_REQUESTS)
            );

            if (dropEvent.compareAndSet(false, true)) {
                logger.warn("Bulk processor is full. Start dropping events.");
            }
        }
    }
```

Reply:
I think that we can also try to modify the bulkProcessor to either have the 500ms time configurable or contain the mechanism you want. So that all future use cases like yours will benefit from it

Reply:
@pgomulka The current code is very stable in my benchmarks under heavy load. I think we will keep this version and I will see what could be improved after the feature freeze. Are you comfortable with this?